### ISSUE TITLE
parse_json, simplify & templates

### ISSUE THREAD

**Issue by dcooley** [knapply/rcppsimdjson#5]
Opening here rather than at `/eddelbuettel` because this is very conceptual at the moment.

I copied your implementation into [jsonify](https://github.com/SymbolixAU/jsonify/blob/issue61/inst/include/jsonify/from_json/parse_json.hpp) and get very similar results*

```r
rl <- readLines("~/Documents/github/sf-city-lots-json/citylots.json")
js <- paste(rl, collapse = "")

microbenchmark::microbenchmark(
  simd = { res <- RcppSimdJson:::.parse_json( js ) },
  jfy = { res <- jsonify:::rcpp_parse_json( js ) },
  jlte = { res <- jsonlite::parse_json( js ) },
  times = 2 
)

# Unit: seconds
# expr       min        lq      mean    median        uq       max neval
# simd  2.267790  2.267790  2.540152  2.540152  2.812514  2.812514     2
#  jfy  2.067401  2.067401  2.541424  2.541424  3.015447  3.015447     2
# jlte 13.438585 13.438585 14.314010 14.314010 15.189435 15.189435     2
```

* identical is FALSE - I need to work out why.

---

I'm also refactoring my [`simplify.hpp`](https://github.com/SymbolixAU/jsonify/blob/issue68/inst/include/jsonify/from_json/simplify/simplify.hpp) code to make it more templated and simpler.

---

What are your thoughts on a templated JSON interface library, {jsonparser}, which both `rcppsimdjson` and `jsonify` can call to do the parsing and simplifying?

**Comment by knapply** [comment]
Sorry, I somehow didn't get a notification for this.

Maybe I'm missing something, but doesn't this create an impossible web of `LinkingTo`? 

```
{jsonify} <=========>  {jsonparser} <========> {RcppSimdJson}`
     |                      |
     |                      |
     |===> {rapidjsonr} <===|
```

I think there's a lot of value in standardization and modularization, but I'm not sure how this could go beyond building a reference interface for some sort of standard. Since that would have to happen as an early step anyways, maybe that's a good target to set? 

Regardless of the direction, I propose the first thing we do is establish that "wish list" of features that a comprehensive JSON R package should have. I have a running list of wants (some of which simply don't exist in R land) that I planned to pin to an issue in `eddelbuettel/RcppSimdJson`. Maybe I should pull the trigger on that sooner rather than later...

**Comment by dcooley** [comment]
My idea for the design is, all my 'simplify' logic, and most of my 'utils', and your `Int64` handling could go into an interface package. Then both {jsonify} and {RcppSimdJson} would call it, and therefore be standardised.

I've written a `wrap()` template to handle specific `rapidjson` objects [here in issue 69](https://github.com/SymbolixAU/jsonify/blob/issue69/inst/include/jsonify/jsonify_types.hpp).

This means my new [`simplify.hpp`](https://github.com/SymbolixAU/jsonify/blob/issue69/inst/include/jsonify/from_json/simplify/simplify.hpp) is (almost*) compelely independent of rapidjson.

*it's still a work in progress, but the final step is to enumerate the various [rapidjson::kTypes](https://github.com/SymbolixAU/jsonify/blob/issue69/inst/include/jsonify/from_json/simplify/simplify.hpp) and it's then free of rapidjson.

so the design is like

```
{rapidjson} ----------|
                      |--- {jsonify}
{jsonparser} ---------
                      |--- {rcppsimdjson}
{simdjson} -----------|

```

**Comment by knapply** [comment]
So you're thinking that `{jsonify}` will be able to link to `{rapidjson}` and `{jsonparser}` in a way that `{rapidjson}` and `{jsonparser}` wouldn't have to depend on each other?

I didn't think that would be possible, but I'm intrigued.

**Comment by dcooley** [comment]
yep. {rapidjson} and {jsonparser} would not touch each other. And neither would {simdjson} and {jsonparser}.

**Comment by knapply** [comment]
> 
> ```
> {rapidjson} ----------|
>                       |--- {jsonify}
> {jsonparser} ---------
>                       |--- {simdjson + rcppsimdjson}
> ```

How would jsonparser not always be the root of the tree? 

Everything with an R interface would need to build its own suite of template specializations so whatever standard is set for jsonparser can hook into.

In your simplify routine, everything is already a list right? So you’re doing stage 1, the json parsing to an equivalent R structure, then stage 2 is simplifying. 

If that’s good enough, then the you’ve already done the generic part, simplifying a list.

What I was trying to suggest with “separating concerns” is not treating these as successive tasks. R is absolutely the bottleneck, so the most efficient approach for simplifying is not materializing anything in R that isn’t already simplified.

**Comment by dcooley** [comment]
> In your simplify routine, everything is already a list right? So you’re doing stage 1, the json parsing to an equivalent R structure, then stage 2 is simplifying.

no I don't think that's quite right. It simplifies as it goes, every time it hits an `Array` it tries to simplify it as it's parsed. 

If the `Array` contains other `Arrays` or `Objects`, then it rescurses into those, parses and tries to simplify, ... and continues parsing / simplifying / parsing / simplifying until it gets back up to the top level object. At which point, all inner-objects are simplifed.

**Comment by knapply** [comment]
Maybe I'm misunderstanding the flow of data.

The reason it seems that way is because all the [`simpify_()*`](https://github.com/SymbolixAU/jsonify/blob/issue69/inst/include/jsonify/from_json/simplify/simplify.hpp?) functions take an `Rcpp::List&`, so the data has already been inserted into an R object before the simplifying begins.

**Comment by dcooley** [comment]
I'm not sure if we're describing the same thing or not 😕   😄 

Yes, all the `simplify` functions take a `List`.

It simplifies the JSON from the inside. So it parses through the whole JSON structure, and when it can't parse any more (i.e., there are no more nested Arrays or Objects), it then simplifies as it comes back out.

**Comment by knapply** [comment]
I didn't catch that it was simplifying on the way out, so I now I'm not sure myself! 

Maybe this will make more sense...

If simplify doesn't copy R objects, calls R less, and can insert data into contiguous structures, it can be much more efficient than the non-simplify approaches. 

I'm pretty sure there's a path to making simplify the fastest approach, regardless of the C++ parsing library.

```r
file_path <- "~/Documents/citylots.json"
js <- readr::read_file(file_path)

# experimental... it's not exactly consistent
p <- RcppSimdJson:::.parser(initial_capacity = object.size(js))

bench::mark(
  simplify = jsonify::from_json(js),
  nosimplify = jsonify::from_json(js, simplify = FALSE),
  parse = jsonify:::rcpp_parse_json(js),
  simdjson = RcppSimdJson:::.parse_json(js, parser_ = p),
  check = FALSE,
  filter_gc = FALSE,
  max_iterations = 3
)
#> # A tibble: 4 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 simplify      6.66s    6.66s     0.150   104.5MB    0.601
#> 2 nosimplify    5.04s    5.04s     0.198    13.1MB    0.595
#> 3 parse         4.99s    4.99s     0.200    13.1MB    0.601
#> 4 simdjson   980.69ms 980.69ms     1.02     13.1MB    0
```

**Comment by knapply** [comment]
@dcooley - did you plan on making a `jsonparser` repo to experiment?

**Comment by dcooley** [comment]
Yes, but first I've been experimenting [here on `issue69` branch](https://github.com/SymbolixAU/jsonify/blob/issue69/inst/include/jsonify/from_json/from_json.hpp) (it's not compiling at the moment, so don't try and install it). I'm replacing anything that's `rapidjson`-specific with a new templated function. 

I see this file then being part of `jsonparser`. Then each of our respective libraries would have to make sure there are functions which implement those templated ones, which I'm doing [here](https://github.com/SymbolixAU/jsonify/blob/issue69/inst/include/jsonify/jsonify_types.hpp).

**Comment by knapply** [comment]
I forgot these multi-column data frame columns even exist...


``` r
# from the {jsonlite} docs
x <- data.frame(driver = c("Bowser", "Peach"), occupation = c("Koopa", "Princess"))
x$vehicle <- data.frame(model = c("Piranha Prowler", "Royal Racer"))
x$vehicle$stats <- data.frame(speed = c(55, 56), weight = c(67, 24), drift = c(35, 32))

json <- jsonlite::toJSON(x)
jsonlite::fromJSON(json)
#>   driver occupation   vehicle.model vehicle.stats.speed vehicle.stats.weight
#> 1 Bowser      Koopa Piranha Prowler                  55                   67
#> 2  Peach   Princess     Royal Racer                  56                   24
#>   vehicle.stats.drift
#> 1                  35
#> 2                  32
jsonify::from_json(json)
#> Registered S3 method overwritten by 'jsonify':
#>   method     from    
#>   print.json jsonlite
#>   driver occupation   vehicle.model vehicle.stats.speed vehicle.stats.weight
#> 1 Bowser      Koopa Piranha Prowler                  55                   67
#> 2  Peach   Princess     Royal Racer                  56                   24
#>   vehicle.stats.drift
#> 1                  35
#> 2                  32

data.table::setDT(jsonify::from_json(json))
#> Warning in data.table::setDT(jsonify::from_json(json)): Some columns are a
#> multi-column type (such as a matrix column): [3]. setDT will retain these
#> columns as-is but subsequent operations like grouping and joining may fail.
#> Please consider as.data.table() instead which will create a new column for each
#> embedded column.
data.table::as.data.table(jsonify::from_json(json))
#>    driver occupation   vehicle.model vehicle.stats.speed vehicle.stats.weight
#> 1: Bowser      Koopa Piranha Prowler                  55                   67
#> 2:  Peach   Princess     Royal Racer                  56                   24
#>    vehicle.stats.drift
#> 1:                  35
#> 2:                  32
tibble::as_tibble(jsonify::from_json(json))
#> # A tibble: 2 x 3
#>   driver occupation vehicle$model   $stats$speed $$weight $$drift
#>   <chr>  <chr>      <chr>                  <int>    <int>   <int>
#> 1 Bowser Koopa      Piranha Prowler           55       67      35
#> 2 Peach  Princess   Royal Racer               56       24      32

dput(jsonify::from_json(json))
#> structure(list(driver = c("Bowser", "Peach"), occupation = c("Koopa", 
#> "Princess"), vehicle = structure(list(model = c("Piranha Prowler", 
#> "Royal Racer"), stats = structure(list(speed = 55:56, weight = c(67L, 
#> 24L), drift = c(35L, 32L)), class = "data.frame", row.names = 1:2)), class = "data.frame", row.names = 1:2)), class = "data.frame", row.names = 1:2)
```


They seem needlessly weird to me and don't seem supported beyond `{base}` (`data.table::setDT()` warns and `data.table::as.data.table()` tosses them out for proper columns)  

Is this something you use (or know why someone else would)?

**Comment by dcooley** [comment]
I do not regularly use them, but when a JSON structure contains nested items like this, I will often split the nested data.frames into separate data.frames, and use them like relational tables (through joins).

**Comment by knapply** [comment]
I've got a  "from scratch" approach that focuses on memory, eliminating branches, and delaying touching R structures as long possible. 

I've got to move a bunch of things around, but hope to have it to rcppsimdjson over the weekend. So far, it's promising....


``` r
.simplify_json <- benchmarking:::.simplify_json

obj <- list(
  replicate(500, sample(c(letters, LETTERS), size = 500, replace = TRUE)),
  replicate(500, sample(c(rep(NA, 100), 1:1000))),
  replicate(500, sample(c(rep(NA, 100), seq(1, 100, by = 0.1)))),
  replicate(500, sample(c(TRUE, FALSE, NA), size = 500, replace = TRUE))
)

json <- jsonify::to_json(obj)

bench::mark(
  simdjson = .simplify_json(json, type_policy = 0, simplify_vector = T, simplify_matrix = T, int64_r_type = 0),
  jsonify = jsonify::from_json(json),
  jsonlite = jsonlite::fromJSON(json),
  filter_gc = FALSE,
  iterations = 5
)

#> # A tibble: 3 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 simdjson     44.5ms   46.3ms     21.5     11.2MB     4.29
#> 2 jsonify     164.2ms  169.3ms      5.82    30.7MB    20.9 
#> 3 jsonlite    489.3ms  517.4ms      1.90    31.7MB     3.42
```

**Comment by dcooley** [comment]
Does your approach keep the nested data.frame structures, or are you "flattening" them in some way?

**Comment by knapply** [comment]
Data frames are what I'm still sorting out... my initial approach didn't make those multi-column things at all, so I'm working on getting that in there as an option. 

It's templated on the user-provided arguments the whole way down, so getting all of that flexibility in from the start seems like it will be _way_ easier than trying to add it later.

**Comment by dcooley** [comment]
cool. Looking forward to seeing what you come up with.
